### ISSUE TITLE
Warn on sub-optimal tuple layout

### ISSUE THREAD

**Issue by gnzlbg** [Manishearth/rust-clippy#1040]
For whatever reason rustc doesn't optimize the layout of tuples to minimize their size.

It would be nice to have a lint that warns about sub-optimal tuple layouts, and that proposes how to improve them. 

The Levenshtein distance can be used to propose a tuple layout with minimum size and minimum number of field reordering by the user. 

It might be worth it to explain that the minimum tuple size can always be obtained by reordering the fields by decreasing size, and also to provide the user with a list of all tuple layouts with the minimum size.  

Consider the following types:

```rust
/// This is what I mean and what makes sense for my program! 
/// Tuples must be zero cost!
type A = (u8, u16, u8, u32);

/// OMG! That tuple is 33% bigger than any of these!
/// Thanks clippy for saving my day! Damn you rustc, damn you!
type B = (u32, u16, u8, u8);
type C = (u16, u8, u8, u32);
type D = (u8, u8, u16, u32);  // min Levenshtein distance
type E = (u32, u8, u8, u16);

fn main () {
 println!("{}", std::mem::size_of::<A>());  // prints 12
 println!("{}", std::mem::size_of::<B>(), 
                std::mem::size_of::<C>(), 
                std::mem::size_of::<D>(),
                std::mem::size_of::<E>());  // prints 8 8 8 8
}
```

```shell
type A = (u8, u16, u8, u32)
         ^~~~~~~~~~~~~~~~~ Sub-optimal tuple layout!
  The memory layout of the tuple is: `(u8, "8-bit padding", u16, u8, "24-bit padding", u32)`
  You can reduce the tuple size by 33% by switching the position of the tuple fields [(2,3)]
  which gives: `(u8, u8, u16, u32)`.
  
  In general, the minimum padding can always be obtained by reordering the tuple-elements 
  by decreasing size, which gives: `(u32, u16, u8, u8)`. 

  Other layouts with minimum size are: `(u16, u8, u8, u32)`, `(u32, u8, u8, u16)`, ...
```

**Comment by llogiq** [comment]
Is this a duplicate of https://github.com/Manishearth/rust-clippy/issues/894

**Comment by gnzlbg** [comment]
@llogiq I think so. The padding issue applies to all data types: structs, enums, tuples, maybe even closures somehow? (even though the user cannot change it?). 

I had only tuples in mind because for structs and unions the layout might depend on things like FFI with C (as opposed to tuples which are "Rust only" so to speak). 

IMO it would be better to close this issue, and just mention tuples in the other one.

**Comment by llogiq** [comment]
Closed as duplicate of #894
